doc/architecture/decisions/0008-add-first-class-citizen-fields-to-applications.md

```diff
@@ -0,0 +1,122 @@
+# 8. Add First-Class Citizen Fields to Applications
+
+Date: 2022-10-31
+
+## Status
+
+Accepted
+
+## Context
+
+When processing an application, there are some situations where we need easy
+access to particular fields within the JSON application object. These include
+(but are not limited to):
+
+- If an application is for a women-only AP
+- If an application is for a PIPE placement
+- The target location
+- The desired start date
+
+There are a number of options we could implement to do this. We could either:
+
+- save these additional fields to the database when an application is updated
+- update the UI code to send these additional fields to the API; or
+- use a [JsonLogic](https://jsonlogic.com) statement to fetch these items as part of the transformation process
+
+All of these approaches come with advantages and drawbacks:
+
+- saving the additional fields will require work at the API side, we would have to query the JSON object whenever it's updated to draw out the relevant fields. We would also have to consider what happens when the schema is updated
+- updating the UI code would further complicate the code at this level. We would have to update our page objects to pull out this data and marshal it into the correct format to send to the API, which would require extra conditionals and complexities. We also have no guarantee that the data would remain in sync.
+- using the JsonLogic statement in the transformer would require work to ensure this was updated every time the schema changed
+
+## Decision
+
+With all these approaches in mind, we have decided to use the transformers to
```

Comment:
Since the frontend already understands the `data` field of an application (and will do for assessments too) I don't think we'd need to return these fields again seperately in the response.  So we'd only need to use the JSON Logic stuff when the API actually needs to make use of a value (e.g. understanding if the application been submitted on a PUT) rather than in the transformer?

Reply:
Hmmm - that's an interesting point actually - maybe we should be using the JSONlogic in the frontend?

Reply:
No, it's just any object at the mo

Reply:
I'm thinking we should probably start typing `data` for our own convinience... if we did would that help with what we're discussing in this ADR? I'm struggling to see how but probably missing something

Reply:
Ah I see, the frontend still "understands" the data though right (i.e. that understanding is dynamically typed but it's still present in the source code) so it would feel a little strange to use normal dot/bracket notation access half of the time then JSON Logic sometimes?

Reply:
I think the motivation here is that we want one source of truth for where these fields will come from, and something that is compatible with schema changes. Keeping the two tightly coupled will ensure we're only changing them in one place when we upgrade the schema.

Reply:
Am I right in thinking that with the updated ADR the answer to this one is that we do not return the first class citizen fields on the API response?